File: frontend/src/types/admin_training.ts
// Types for the ML Training Operator Console — Phase 1

export type ProbeStatus = 'pass' | 'warn' | 'fail';

export interface ProbeResult {
  name: string;
  label: string;
  status: ProbeStatus;
  message: string;
  value?: unknown;
  remediation?: string | null;
}

export interface PreflightReport {
  can_launch: boolean;
  probes: ProbeResult[];
  checked_at: string;
}

export interface TrainingConfigOverride {
  model_version?: string | null;
  n_symbols?: number | null;
  lookback_years?: number | null;
  xgboost_trials?: number | null;
  gru_trials?: number | null;
}

export interface LaunchRequest {
  reason: string;
  override_schedule_warning?: boolean;
  config?: TrainingConfigOverride;
  feedback_weights_path?: string | null;
}

export interface FeedbackBundleInfo {
  bundle_path:        string;
  sha256:             string;
  /** Number of unique (instrument_key, signal_date) rows in the parquet. */
  row_count:          number;
  /** Total closed trades aggregated into this bundle (≥ row_count). */
  total_raw_outcomes: number;
  window_start:       string;
  window_end:         string;
  created_at:         string;
  weight_mean:        number;
  weight_std:         number;
  weight_p5:          number;
  weight_p50:         number;
  weight_p95:         number;
  histogram_bins:     number[];
  histogram_counts:   number[];
  top_symbols:        Array<{ symbol: string; mean_weight: number; n_outcomes: number }>;
}

export interface FeedbackBundlesListResponse {
  bundles: FeedbackBundleInfo[];
  total: number;
}

export interface LaunchResponse {
  run_id: string;
  launched_at: string;
  checkpoint_dir: string;
  message: string;
}

export type RunStatus = 'running' | 'completed' | 'failed' | 'unknown';

export interface RunSummary {
  run_id: string;
  status: RunStatus;
  started_at?: string | null;
  completed_at?: string | null;
  duration_s?: number | null;
  mlflow_run_id?: string | null;
  model_version?: string | null;
  steps_completed: string[];
  source: string;
  error_summary?: string | null;
}

export interface RunsListResponse {
  runs: RunSummary[];
  total: number;
}

export interface ActiveRunResponse {
  active: boolean;
  run_id?: string | null;
  launched_at?: string | null;
  pid?: number | null;
}

// WebSocket frame types (server → client)

export interface RunEventFrame {
  type: 'run_event';
  data: RunLogEntry;
}

export interface RunCompleteFrame {
  type: 'run_complete';
  exit_code: number;
  success: boolean;
}

export interface ConnectedFrame {
  type: 'connected';
  run_id: string;
}

export interface ErrorFrame {
  type: 'error';
  code: string;
  message?: string;
}

export interface PongFrame {
  type: 'pong';
  ts: string;
}

export type TrainingWsFrame =
  | RunEventFrame
  | RunCompleteFrame
  | ConnectedFrame
  | ErrorFrame
  | PongFrame;

// Structured log entry from run_log.ndjson

export interface RunLogEntry {
  ts: string;
  event: string;
  // run_start
  run_id?: string;
  mlflow_run_id?: string;
  model_version?: string;
  resumed?: boolean;
  // step_complete
  step?: string;
  step_num?: number;
  duration_s?: number;
  metrics?: Record<string, number>;
  // run_failed / run_finished
  promotion_report_path?: string | null;
  // c2_report_generated
  status?: string;
  report_path?: string | null;
  // f1_event_backtest_complete
  agreement_status?: string;
  divergence_pct?: number | null;
  mean_ann_sharpe?: number | null;
  total_fills?: number | null;
}

// UI state

export type TrainingTab = 'preflight' | 'launch' | 'live' | 'history' | 'feedback';

export const PIPELINE_STEPS: Array<{ key: string; label: string }> = [
  { key: 'step_1_symbols',   label: 'Symbol Selection'       },
  { key: 'step_2_features',  label: 'Feature Computation'    },
  { key: 'step_3_targets',   label: 'Target Generation'      },
  { key: 'step_4_splits',    label: 'CV Plan (CPCV)'         },
  { key: 'step_5_xgboost',   label: 'XGBoost Training'       },
  { key: 'step_6_gru',       label: 'GRU Training'           },
  { key: 'step_7_ensemble',  label: 'Ensemble Optimisation'  },
  { key: 'step_8_evaluation','label': 'Evaluation'           },
  { key: 'step_9_onnx',      label: 'ONNX Export'            },
  { key: 'step_10_registry', label: 'Model Registry'         },
];

File: backend/app/schemas/admin_training.py
"""
Pydantic schemas for the ML Training Operator Console — Phase 1.
"""
from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

_NO_NAMESPACE = ConfigDict(protected_namespaces=())


class ProbeStatus(str, Enum):
    PASS = "pass"
    WARN = "warn"
    FAIL = "fail"


class ProbeResult(BaseModel):
    name: str
    label: str
    status: ProbeStatus
    message: str
    value: Any = None
    remediation: str | None = None


class PreflightReport(BaseModel):
    can_launch: bool
    probes: list[ProbeResult]
    checked_at: str


class TrainingConfigOverride(BaseModel):
    model_config = ConfigDict(extra="forbid", protected_namespaces=())

    model_version: str | None = Field(None, pattern=r"^\d+\.\d+\.\d+$")
    n_symbols: int | None = Field(None, ge=10, le=5000)
    lookback_years: int | None = Field(None, ge=1, le=20)
    xgboost_trials: int | None = Field(None, ge=1, le=500)
    gru_trials: int | None = Field(None, ge=1, le=100)


class LaunchMode(str, Enum):
    FRESH  = "fresh"   # --fresh: discard any existing checkpoint, start from scratch
    RESUME = "resume"  # no --fresh: auto-resume from the existing checkpoint


class LaunchRequest(BaseModel):
    reason: str = Field(min_length=5, max_length=500)
    mode: LaunchMode = LaunchMode.FRESH
    override_schedule_warning: bool = False
    # config overrides and feedback_weights_path are only honoured in FRESH mode;
    # RESUME ignores them — the checkpoint's locked config governs the resumed run.
    config: TrainingConfigOverride = Field(default_factory=TrainingConfigOverride)
    feedback_weights_path: str | None = None


class CheckpointStatus(BaseModel):
    """Current on-disk checkpoint state — used by the Launch form to decide mode."""

    model_config = ConfigDict(protected_namespaces=())

    exists: bool
    resumable: bool            # exists AND incomplete AND schema-compatible
    run_id: str | None = None
    started_at: str | None = None
    model_version: str | None = None
    schema_version: int | None = None
    completed_steps: list[str] = Field(default_factory=list)
    next_step: str | None = None
    # GRU sub-C state — populated when step_6_gru is the next step and a
    # per-epoch training_state.json exists (i.e. GRU was interrupted mid-run).
    gru_last_epoch: int | None = None


class LaunchResponse(BaseModel):
    run_id: str
    launched_at: str
    checkpoint_dir: str
    message: str


class RunStatus(str, Enum):
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    UNKNOWN = "unknown"


class RunSummary(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    run_id: str
    status: RunStatus
    started_at: str | None = None
    completed_at: str | None = None
    duration_s: float | None = None
    mlflow_run_id: str | None = None
    model_version: str | None = None
    steps_completed: list[str] = Field(default_factory=list)
    source: str = "unknown"
    error_summary: str | None = None


class RunsListResponse(BaseModel):
    runs: list[RunSummary]
    total: int


class ActiveRunResponse(BaseModel):
    active: bool
    run_id: str | None = None
    launched_at: str | None = None
    pid: int | None = None


# ── Phase 2: Feedback weight bundle schemas ────────────────────────────────────

class FeedbackBundleInfo(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    bundle_path:        str
    sha256:             str
    row_count:          int   # unique (instrument_key, signal_date) pairs in the parquet
    total_raw_outcomes: int   # sum of outcome_count — total closed trades aggregated
    window_start:       str
    window_end:         str
    created_at:         str
    weight_mean:        float
    weight_std:         float
    weight_p5:          float
    weight_p50:         float
    weight_p95:         float
    histogram_bins:     list[float]
    histogram_counts:   list[int]
    top_symbols:        list[dict[str, Any]]


class FeedbackBundlesListResponse(BaseModel):
    bundles: list[FeedbackBundleInfo]
    total: int
